File: Language_Essentials/java/Q3SetOps.java
// Set Operations - TCS NQT
import java.util.HashSet;
import java.util.TreeSet;

public class Q3SetOps {
    public static void main(String[] args) {
        // --- 1. TreeSet (Stores unique elements in sorted order) ---
        System.out.println("--- TreeSet ---");
        TreeSet<Integer> ts = new TreeSet<>();
        
        // Insert elements
        ts.add(10);
        ts.add(5);
        ts.add(20);
        ts.add(10); // Duplicate, will not be added
        
        System.out.print("TreeSet elements: ");
        for (int x : ts) {
            System.out.print(x + " "); // Output: 5 10 20
        }
        System.out.println();
        
        // Check if element exists
        if (ts.contains(10)) {
            System.out.println("10 is present in the set.");
        }
        
        // Erase element
        ts.remove(5);
        System.out.println("After erasing 5, size is: " + ts.size());
        
        // --- 2. HashSet (Faster O(1) time, checks for duplicates without sorting) ---
        System.out.println("\n--- HashSet ---");
        HashSet<Integer> hs = new HashSet<>();
        hs.add(3);
        hs.add(1);
        hs.add(3); // Ignored
        
        System.out.print("HashSet elements: ");
        for (int x : hs) {
            System.out.print(x + " "); 
        }
        System.out.println();
    }
}

File: Language_Essentials/python/q3_set_ops.py
# Set Operations - TCS NQT

def set_ops():
    # --- 1. Ordered Set (Python's set is unordered, but preserves insertion order in Python 3.7+) ---
    print("--- Set Operations ---")
    s = set()
    
    # Insert elements
    s.add(10)
    s.add(5)
    s.add(20)
    s.add(10) # Duplicate, will not be added
    
    print("Set elements:")
    for x in s:
        print(x, end=" ") # Order is not guaranteed to be sorted!
    print()
    
    # Check if element exists (O(1) time)
    if 10 in s:
        print("10 is present in the set.")
        
    # Erase element
    if 5 in s:
        s.remove(5)
        
    print(f"After erasing 5, size is: {len(s)}")
    
    # --- 2. Set operations ---
    print("\n--- Additional Set Ops ---")
    s2 = {1, 3, 20}
    
    print("Intersection:", s.intersection(s2))
    print("Union:", s.union(s2))
    print("Difference (s - s2):", s.difference(s2))


if __name__ == "__main__":
    set_ops()
